File: 2015/ruby/day15.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end

data = File.read('day15.txt')

sample = <<~TXT
Butterscotch: capacity -1, durability -2, flavor 6, texture 3, calories 8
Cinnamon: capacity 2, durability 3, flavor -2, texture -1, calories 3
TXT
def parse_line(line)
  name, rest = line.split(': ')
  capacity, durability, flavor, texture, calories = rest.scan(/[\-\d]+/).map(&:to_i)
  {name: name, capacity: capacity, durability: durability, flavor: flavor, texture: texture, calories: calories}
end

def amts(buckets, total)
  return [[total]] if buckets == 1

  rv = []
  (0..total).each do |n|
    amts(buckets - 1, total - n).each do |old|
      rv.push(old + [n])
    end
  end
  rv
end

def score(ingredients_added)
  score = 0
  values = %i(capacity durability flavor texture).map do |t|
    ingredients_added.map do |(amt, ingredient)|
      amt * ingredient[t]
    end.sum
  end
  values.reduce(1) {|acc, v| acc * [v, 0].max}
end

def highest_score(data)
  ingredients = data.lines.map {|l| parse_line(l)}
  amts = amts(ingredients.size, 100)
  amts.map do |amt|
    score(amt.zip(ingredients))
  end.max
end
assert_equal([[100]], amts(1, 100))
assert_equal([[0,1], [1,0]].sort, amts(2, 1).sort)
assert_equal([[0,2], [1,1], [2,0]].sort, amts(2, 2).sort)
assert_equal([[2,0,0], [1,1,0], [1,0,1], [0, 2, 0], [0, 1, 1], [0, 0, 2]].sort, amts(3, 2).sort)

assert_equal({name: 'Butterscotch', capacity: -1, durability: -2, flavor: 6, texture: 3, calories: 8}, parse_line(sample.lines[0]))

assert_equal(62842880, highest_score(sample))
puts "Part 1: #{highest_score(data)}"

File: 2015/go/day15/day15.go
package day15

import (
	"fmt"
	"strings"
)

type Ingredient struct {
	Name                                            string
	Capacity, Durability, Flavor, Texture, Calories int
}

func ParseLine(line string) Ingredient {
	var name string
	var capacity, durability, flavor, texture, calories int
	fmt.Sscanf(line, "%s capacity %d, durability %d, flavor %d, texture %d, calories %d",
		&name, &capacity, &durability, &flavor, &texture, &calories)

	return Ingredient{strings.TrimSuffix(name, ":"), capacity, durability, flavor, texture, calories}
}

func Amounts(buckets int, total int) [][]int {
	if buckets == 1 {
		return [][]int{{total}}
	}
	rv := make([][]int, 0)
	for n := 0; n <= total; n++ {
		for _, old := range Amounts(buckets-1, total-n) {
			rv = append(rv, append(old, n))
		}
	}
	return rv
}
func Score(amounts []int, ingredients []Ingredient, requiredCalories int) int {
	type Added struct {
		amt        int
		ingredient Ingredient
	}
	ingredientsAdded := make([]Added, len(amounts))
	for idx, amt := range amounts {
		ingredientsAdded[idx] = Added{amt, ingredients[idx]}
	}
	if requiredCalories > 0 {
		sum := 0
		for _, added := range ingredientsAdded {
			sum += added.amt * added.ingredient.Calories
		}
		if sum != requiredCalories {
			return 0
		}
	}
	values := make([]int, 4)
	for _, added := range ingredientsAdded {
		values[0] += added.amt * added.ingredient.Capacity
		values[1] += added.amt * added.ingredient.Durability
		values[2] += added.amt * added.ingredient.Flavor
		values[3] += added.amt * added.ingredient.Texture
	}
	sum := 1
	for _, v := range values {
		if v <= 0 {
			return 0
		}
		sum *= v
	}
	return sum
}

func HighestScore(input []string, requiredCalories int) int {
	ingredients := make([]Ingredient, len(input))
	for idx, line := range input {
		ingredients[idx] = ParseLine(line)
	}
	amts := Amounts(len(ingredients), 100)
	max := -1
	for _, amt := range amts {
		s := Score(amt, ingredients, requiredCalories)
		if s > max {
			max = s
		}
	}
	return max
}
